broker/broker_utils/broker_utils/data_utils.py

```diff
@@ -6,24 +6,91 @@
 
 from astropy.time import Time
 import fastavro
+from io import BytesIO
+import json
 import numpy as np
 import pandas as pd
-from typing import Tuple
+from typing import Optional, Tuple, Union
 
 
-def alert_avro_to_dict(filename: str) -> dict:
-    """Load an alert Avro file to a dictionary.
+def decode_alert(
+    alert_avro: Union[str, bytes],
+    return_as: str = 'dict',
+    schema_map: Optional[dict] = None,
+    drop_cutouts: bool = False,
+) -> Union[dict, pd.DataFrame]:
+    """Load an alert Avro and return in requested format.
+
+    Wraps `alert_avro_to_dict()` and `alert_dict_to_dataframe()`.
+
+    Args:
+        alert_avro:   Either the path of Avro file to load, or
+            the bytes encoding the Avro-formated alert.
+        return_as: Format the alert will be returned in.
+        schema_map: Mapping between survey schema and broker's generic schema.
+            Required if `return_as='df'` or `drop_cutouts=True`.
+        drop_cutouts: Whether to drop or return the cutouts (stamps).
+                      If `return_as='df'` the cutouts are always dropped.
+
+    Returns:
+        alert packet in requested format
+    """
+    alert_dict = alert_avro_to_dict(alert_avro)
+
+    if return_as == "dict":
+        if drop_cutouts:
+            return _drop_cutouts(alert_dict, schema_map)
```

Comment:
@djperrefort Codacy complains about this line, saying:

> "Argument 2 to "_drop_cutouts" has incompatible type "Optional[Dict[Any, Any]]"; expected "Dict[Any, Any]" (error)".

I doesn't seem to know about `typing.Optional` Any ideas on how to fix that? I tried a Google search, but didn't get very far.

Apparently the error is too minor to be flagged here, but I see it on the [Codacy GUI](https://app.codacy.com/gh/mwvgroup/Pitt-Google-Broker/pullRequest?_ga=2.222584475.1032047264.1630857686-1452767873.1630601139)

Reply:
@troyraen The error message is correct. The inspection tool is telling you that the function signatures are not compatible.

The `_drop_cutouts` function is type hinted to imply that the second argument (`schema_map`) is **not** optional. However,  the signature of the `decode_alert` function indicates that `schema_map` may be `None`. 

Type hinting is used to help prevent bugs that can creep into code due to duck-typing. In this case, either the type hints are incorrect, or the the functions are being called in a way that violates the way the function is supposed to be called.

Reply:
@djperrefort Ah, thank you. Can you help me code this correctly? `schema_map` is __not__ required if `drop_cutouts=False`. So I want it to be optional in `decode_alert()`, even though it is required in `_drop_cutouts()`. How do I write that correctly?

Reply:
The simplest way is to LBYL and check the `schema_map` argument is not `None` before calling `_drop_cutouts`. If it is `None` then raise an error.

The more complex way is to add multiple type hint signatures:

```python

from typing import overload

@overload
def decode_alert(
    alert_avro: Union[str, bytes],
    return_as: str,
    schema_map: dict,
    drop_cutouts: bool,
) -> Union[dict, pd.DataFrame]:
    ...


@overload
def decode_alert(
    alert_avro: Union[str, bytes],
    return_as: str,
    schema_map: None,
    drop_cutouts: bool,
) -> Union[dict, pd.DataFrame]:
    ...


def decode_alert(alert_avro, return_as='dict', schema_map=None, drop_cutouts=False):
    ...  # source code here
```

Which one you use depends on how important the type inspections are to your use case. I suggest using the former for internal code and the latter for code you expect to ship out as an API for use by end users.

Reply:
@djperrefort I added these LBYL checks as you suggested, but it seems overly complicated to me. I have to do 2 checks here, but the function that **actually uses** the `schema_map` already checks the type and raises an error if it's not as expected. 
In other words, this adds 2 redundant checks, requiring 6 new lines of code.  Is this really how I should do it? 

There are other similar situations in the code, and it seems dumb to do redundant checks in every single place. Is there not a good way to let the function that actually uses the `schema_map` handle the type check by itself?

Reply:
> Is there not a good way to let the function that actually uses the schema_map handle the type check by itself?

I'm not sure I understand the question. That should be what the overload example above is doing.

At the end of the day it's just a type check. It can be ignored and revisited as part of our gradual efforts to address codacy inspections in the future.

Reply:
> That should be what the overload example above is doing.

No, `decode_alert()` is not the function that actually uses the `schema_map`. 
`drop_cutouts()` uses the `schema_map`, and it **already** checks to make sure that it is a valid type. 

> I'm not sure I understand the question.

Is there a best-practice way to allow functions that do not actually use the `schema_map` (they just pass it on to another function) to be ignorant of the type? It seems so much simpler to let them pass `schema_map=None` because `drop_cutouts` will handle the check.

> At the end of the day it's just a type check. It can be ignored and revisited as part of our gradual efforts to address codacy inspections in the future.

Yes, I don't really care about these 6 lines. Really, I'm thinking about all of my code going forward and trying to understand best practices.

Reply:
I have since realized that it's probably better to take `schema_map` out of the `decode_alert` signature, and replace it with `**kwargs` which can then be passed on to `_drop_cutouts`.